From: Andreas Herrmann <andreas.herrmann3@amd.com>
Subject: x86, amd: Support L3 Cache Partitioning on AMD family 0x15 CPUs
References: fate#309727
Patch-mainline: v2.6.39-rc1

Signed-off-by: Thomas Renninger <trenn@suse.de>

L3 Cache Partitioning allows selecting which of the 4 L3 subcaches can be used
for evictions by the L2 cache of each compute unit. By writing a 4-bit
hexadecimal mask into the the sysfs file
/sys/devices/system/cpu/cpuX/cache/index3/subcaches, the user can set the
enabled subcaches for a CPU.

The settings are directly read from and written to the hardware, so there is no
way to have contradicting settings for two CPUs belonging to the same compute
unit. Writing will always overwrite any previous setting for a compute unit.

Signed-off-by: Hans Rosenfeld <hans.rosenfeld@amd.com>
Cc: <Andreas.Herrmann3@amd.com>
LKML-Reference: <1297098639-431383-1-git-send-email-hans.rosenfeld@amd.com>
[ -v3: minor style fixes ]
Signed-off-by: Ingo Molnar <mingo@elte.hu>

Signed-off-by: Hans Rosenfeld <hans.rosenfeld@amd.com>
Automatically created from "patches.arch/0124-x86-amd-Support-L3-Cache-Partitioning-on-AMD-family-.patch0124-x86-amd-Support-L3-Cache-Partitioning-on-AMD-family-.patch" by xen-port-patches.py

--- head-2011-05-23.orig/arch/x86/kernel/amd_nb.c	2011-04-13 13:47:56.000000000 +0200
+++ head-2011-05-23/arch/x86/kernel/amd_nb.c	2011-04-15 09:14:01.000000000 +0200
@@ -123,6 +123,7 @@ bool __init early_is_amd_nb(u32 device)
 	return false;
 }
 
+#ifndef CONFIG_XEN
 int amd_get_subcaches(int cpu)
 {
 	struct pci_dev *link = node_to_amd_nb(amd_get_nb_id(cpu))->link;
@@ -181,6 +182,7 @@ int amd_set_subcaches(int cpu, int mask)
 
 	return 0;
 }
+#endif
 
 static int amd_cache_gart(void)
 {
